### PR TITLE
MP#8 rating and review api curl

### PR DESCRIPTION
**Pull request by zijianguan0204**
DROP TABLE IF EXISTS productreviews;
CREATE TABLE productreviews(
    id INT NOT NULL PRIMARY KEY AUTO_INCREMENT,
    productID INT NOT NULL,
    userID INT NOT NULL,
    userName VARCHAR(100) NOT NULL,
    rating INT NOT NULL,
    review VARCHAR(1000) NOT NULL,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);

### CODE REVIEW

**Comment by KunSu** [comment]
> Please setup your database using the following code, test cases are included
> 
> CREATE TABLE `productreviews` (
> `id` int(10) UNSIGNED NOT NULL,
> `productID` int(11) NOT NULL,
> `userID` int(11) NOT NULL,
> `rating` int(11) NOT NULL,
> `review` varchar(1000) NOT NULL,
> `created_at` timestamp NOT NULL DEFAULT current_timestamp()
> ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
> 
> INSERT INTO `productreviews` (`id`, `productID`, `userID`, `rating`, `review`, `created_at`) VALUES
> (1, 12, 123, 4, 'test', '2019-12-09 00:38:59'),
> (2, 234, 234, 5, 'testing 1', '2019-12-09 01:29:13'),
> (3, 41, 321, 3, 'Good', '2019-12-09 03:58:00'),
> (4, 41, 321, 3, 'Good', '2019-12-09 04:09:19'),
> (20, 0, 1, 3, 'final test', '2019-12-10 00:14:42'),
> (21, 0, 1, 3, 'tyang test', '2019-12-10 00:21:39'),
> (22, 0, 1, 3, 'hello tyang', '2019-12-10 00:32:07'),
> (23, 0, 1, 1, 'tyang test1', '2019-12-10 03:09:43'),
> (24, 0, 1, 2, 'testing ....', '2019-12-10 08:57:24'),
> (25, 0, 1, 3, 'hellotes', '2019-12-10 09:01:20'),
> (26, 0, 1, 4, 'testing', '2019-12-10 09:03:39'),
> (27, 41, 1, 3, 'tesing', '2019-12-10 09:05:49'),
> (28, 41, 1, 5, 'real testing', '2019-12-10 09:06:08'),
> (29, 41, 1, 4, 'for real testing', '2019-12-10 09:06:26'),
> (30, 25, 1, 4, 'tyang testing', '2019-12-10 09:07:38'),
> (31, 25, 1, 3, 'haha', '2019-12-10 09:08:03'),
> (32, 27, 1, 3, 'web dev tyang test', '2019-12-10 09:10:51'),
> (33, 11, 1, 3, 'bubble tea test lijuan', '2019-12-10 09:14:00'),
> (34, 17, 1, 4, 'tsting', '2019-12-10 09:15:51');
> 
> ALTER TABLE `productreviews`
> ADD PRIMARY KEY (`id`);
> 
> ALTER TABLE `productreviews`
> MODIFY `id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=35;
> COMMIT;

First of all, do not hardcode `id`
And why there is a productID with 234?
And userID with 123?
Plz enforce foreign key in your table creation. 
For `created_at`, you can use `NOW()` instead of hardcode it.
Delete all unnecessary data.
Not ALTER TABLE at creation state.

File: comment&rating.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
+<!DOCTYPE html>
```
Comment (KunSu) [inline comment&rating.php]: Do not put special charact in the file name

File: comment&rating.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  <html> 
+    <div class = "comment">
+      Leave your comment and rating below!
```
Comment (KunSu) [inline comment&rating.php]: Use HTML tag plz

File: comment&rating.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+      <?php
+        function test_input($data) {
```
Comment (KunSu) [inline comment&rating.php]: test input for what?

File: comment&rating.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+          $data = htmlspecialchars($data);
+          return $data;
+        }
+
+        $rateErr = $commentErr= "";
+        $rate = $comment = "";
+
+        if ($_SERVER["REQUEST_METHOD"] == "POST") {
+
+          if (empty($_POST["comment"])) {
```
Comment (KunSu) [inline comment&rating.php]: trim the input plz. ex comment "           " is not empty

File: comment&rating.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            $rate = test_input($_POST["rate"]);
+          }
+
+          if(!empty($_POST["rate"]) && !empty($_POST["comment"])){
+            echo $comment . " and " . $rate;
+            header("Location: ../../rate.php?name=" . $include_name. "&rating=" . $rate. "&comment=" . $comment);
+          }
+
+        }
+
```
Comment (KunSu) [inline comment&rating.php]: delete empty lines

File: comment&rating.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        
+
+      ?>
+
+      <form method="post" action="<?php echo htmlspecialchars($_SERVER["PHP_SELF"]);?>">
+        <center>
+          <textarea name="comment" rows="4" cols="80"><?php echo $comment;?></textarea>
```
Comment (KunSu) [inline comment&rating.php]: use CSS class instead of hardcoding

File: comment&rating.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        <input type="radio" name="rate" <?php if (isset($rate) && $rate=="4") $rating = 4;?> value="4">4
+        <input type="radio" name="rate" <?php if (isset($rate) && $rate=="5") $rating = 5;?> value="5">5
+
+        <span class="error">* <?php echo $rateErr;?></span>
+        <br><br>
+        <input type="submit" name="submit" value="Submit">
+      </form>
+
+      <br>
+       <div class = "displayComment">
```
Comment (KunSu) [inline comment&rating.php]: put the whole displayComment inside the displayRatingReviews

File: companies/junyanwang/A Frozen SingAlong Celebration.php [hunk 8]
```diff
@@ -87,6 +87,11 @@
          </div> <!-- /entry -->	      
 
       </div> <!-- /portfolio-content -->
+
+    <?php
+    $include_name = basename(__FILE__, '.php');
+    include "$path/comment&rating.php"
```
Comment (KunSu) [inline companies/junyanwang/A Frozen SingAlong Celebration.php]: pass in the $include_name to a function inside php. NO ONE knows where when and why you use it.

File: companies/ruzhang/Mario Kart 8 Deluxe.php [hunk 9]
```diff
@@ -87,6 +87,10 @@
          </div> <!-- /entry -->
 
       </div> <!-- /portfolio-content -->
+      <?php
```
Comment (KunSu) [inline companies/ruzhang/Mario Kart 8 Deluxe.php:97]: Put this code block in the SAME POSITION PLZ. Every single one of them.

File: companies/zijian/User Management System.php [hunk 10]
```diff
@@ -1,8 +1,8 @@
 <?php
-   $path = "../../";
```
Comment (KunSu) [inline companies/zijian/User Management System.php]: No need to change the spacing.

File: companies/zijian/User Management System.php [hunk 11]
```diff
@@ -17,81 +17,87 @@
 
 <body>
 
-   <!-- Header
```
Comment (KunSu) [inline companies/zijian/User Management System.php:22]: I do not know where did you change your code.

Comment (KunSu) [reply companies/zijian/User Management System.php:22]: And why you need to change your code.

File: cookies.php [hunk 12]
```diff
@@ -1,4 +1,5 @@
 <?php
+ob_start();
```
Comment (KunSu) [inline cookies.php:2]: what is this?

File: displayRatingReviews.php [hunk 13]
```diff
@@ -0,0 +1,45 @@
+<?php
+    require_once "config.php";
```
Comment (KunSu) [inline displayRatingReviews.php:2]: Make it as a function and pass in the value you need as parameter plz.

File: users.json [hunk 14]
```diff
@@ -1 +1 @@
-{"0":"kunsu","username":"kunsu","1":"kunsu.903@gmail.com","email":"kunsu.903@gmail.com","2":"1234567890","homephone":"1234567890"}
```
Comment (KunSu) [inline users.json:1]: Delete this change

**Review by KunSu — CHANGES_REQUESTED** [review]
Need a second visit

**Comment by KunSu** [comment]
Plz make sure the SQL is runnable for multiple runs

**Comment by KunSu** [comment]
What is the red thing?
image

**Comment by KunSu** [comment]
image

???

**Comment by KunSu** [comment]
Plz at least do some handling on "long" test.
image

**Comment by KunSu** [comment]
Top Five Review and Rating is missing @zijianguan0204 @rzhang85

File: cookies.php (hunk 12 above)
Comment (rzhang85) [reply cookies.php:2]: Need this for redirect.

File: companies/ruzhang/Mario Kart 8 Deluxe.php (hunk 9 above)
Comment (rzhang85) [reply companies/ruzhang/Mario Kart 8 Deluxe.php:97]: I don't get this one.

File: comment&rating.php (hunk 3 above)
Comment (rzhang85) [reply comment&rating.php]: validation on input data

**Review by KunSu — APPROVED** [review]

**Pull request merged**